snuba/query/snql/parser.py

```diff
@@ -1537,29 +1543,67 @@ def parse_snql_query(
         with sentry_sdk.start_span(op="processor", description="treeify_conditions"):
             _post_process(query, [_treeify_or_and_conditions], settings)
 
+        # dummy settings and timer dont matter
+        dummy_settings = HTTPQuerySettings()
+        dummy_timer = Timer("snql_pipeline")
+        res = PostProcessAndValidateQuery().execute(
+            QueryPipelineResult(
+                data=(query, dataset, settings, custom_processing),
+                error=None,
+                query_settings=dummy_settings,
+                timer=dummy_timer,
+            )
+        )
+        if res.error:
+            raise res.error
+        assert res.data  # since theres no res.error, data is guaranteed
+        return res.data
+    except InvalidQueryException:
+        raise
+    except Exception:
+        raise PostProcessingError(query)
+
+
+class PostProcessAndValidateQuery(
+    QueryPipelineStage[
+        tuple[
+            LogicalQuery | CompositeQuery[LogicalDataSource],
+            Dataset,
+            QuerySettings | None,
+            CustomProcessors | None,
+        ],
+        LogicalQuery | CompositeQuery[LogicalDataSource],
+    ]
+):
+    def _process_data(
+        self,
+        pipe_input: QueryPipelineData[
+            tuple[
+                LogicalQuery | CompositeQuery[LogicalDataSource],
+                Dataset,
+                QuerySettings | None,
+                CustomProcessors | None,
```

Comment:
I would leave CustomProcessors out of this stage, make a separate stage with them

Reply:
Can I leave this for another time bc to factor it out into its own stage would require quite a few changes and I just want to get this done

Reply:
why would it require quite a few changes? custom processing is barely used in the codebase

Reply:
When I took custom processing out of the stage, this puts it before `replace_time_condition`, it turns out that it actually does things that need to come before. So this breaks things. I see my options as:

1. leave custom processing as part of the stage as it was
2. move custom processing before the stage instead of after to see it it works
3. split the stage into 3: post_proc1 -> custom_proc -> post_proc2

I would choose option 1, I see it being the most quick, easy, and safe change. With the tradeoff being the stages are less independent.